tests/binary_agreement.rs

```diff
@@ -27,31 +27,139 @@ extern crate threshold_crypto as crypto;
 
 mod network;
 
+use std::collections::BTreeMap;
 use std::iter::once;
 use std::sync::Arc;
 
 use rand::Rng;
 
-use hbbft::binary_agreement::BinaryAgreement;
-use hbbft::NetworkInfo;
+use hbbft::binary_agreement::sbv_broadcast;
+use hbbft::binary_agreement::{BinaryAgreement, Message, MessageContent};
+use hbbft::{NetworkInfo, Target, TargetedMessage};
 
-use network::{Adversary, MessageScheduler, NodeId, SilentAdversary, TestNetwork, TestNode};
+use network::{
+    Adversary, MessageScheduler, MessageWithSender, NodeId, SilentAdversary, TestNetwork, TestNode,
+};
+
+/// An adversary for the reordering attack.
+/// Described here: https://github.com/amiller/HoneyBadgerBFT/issues/59#issue-310368284
+/// A0 is the first third of nodes, A1 is the second third, and the rest are B.
+#[derive(Default)]
+pub struct ReorderingAdversary {
+    /// Node ids seen by the adversary, but not under control of the adversary.
+    known_node_ids: Vec<NodeId>,
+    /// Node ids under control of adversary.
+    known_adversarial_ids: Vec<NodeId>,
+    /// The most recent epoch we've seen.
+    current_epoch: u32,
+    /// If we've sent the one-time messages for the attack.
+    sent_attack: bool,
+}
+
+impl Adversary<BinaryAgreement<NodeId>> for ReorderingAdversary {
+    fn init(
+        &mut self,
+        all_nodes: &BTreeMap<NodeId, TestNode<BinaryAgreement<NodeId>>>,
+        nodes: &BTreeMap<NodeId, Arc<NetworkInfo<NodeId>>>,
+    ) {
+        self.known_adversarial_ids = nodes.keys().cloned().collect();
+        self.known_node_ids = all_nodes.keys().cloned().collect();
+    }
+
+    fn pick_node(&self, nodes: &BTreeMap<NodeId, TestNode<BinaryAgreement<NodeId>>>) -> NodeId {
+        // Prioritizes A0, as required by the attack.
+        MessageScheduler::First.pick_node(nodes)
+    }
+
+    fn push_message(&mut self, _: NodeId, msg: TargetedMessage<Message, NodeId>) {
+        if self.current_epoch < msg.message.epoch {
+            self.current_epoch = msg.message.epoch;
+            // Reset for the next epoch.
+            self.sent_attack = false;
+        }
+    }
+
+    fn step(&mut self) -> Vec<MessageWithSender<BinaryAgreement<NodeId>>> {
+        let mut messages = Vec::new();
+        if !self.sent_attack {
+            self.sent_attack = true;
+            for &our_node_id in &self.known_adversarial_ids {
+                let f = self.known_node_ids.len() / 3;
```

Comment:
`self.known_node_ids` doesn't change inside the loop, I think? Then computing `f` could be moved outside of the loop.